---
title: "Allele call for ASP qPCR typing"
author: "Dmitrii Romaniuk"
date: 'Jan 24, 2019'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

## Tool to make allele call for qPCR with allele specific primers and gene specific probes.
The idea is to use hydrolysis probes to multiplex and speed up PCR with allele-specific primers based genotyping method, usually made in two steps: PCR and electrophoresis. The program is dedicated for particular qPCR setting. Two allele-specific primers are used, so two separate wells needed for each polymorphism typing. To be able to ‘synchronize’ the said wells the control gene qPCR is also included into each well. The typing qPCR reaction could be negative if there was no needed allele for the primer, so the control gene qPCR also helps to check for pipetting errors – the control qPCR always positive, its Cq depends on sample concentration.

The method and one of plausible application were described in Romaniuk et al., Front. Immunology 2019 (in press)

# Make sure your RDML file has all targets named (no template control included).
# Example: "Target1_wild", "Target1_mut" (note underscore).
# Two cells with same sample name, target name (before underscore) will be 
# compared and call will be made for each target state written after underscore.
  
# The tool was checked for RDML v1.1 files made with BioRad CFX manager v3.1.

Press Ctrl+Shift+Enter to run needed cell below.

OPTIONAL: Install packages.
```{r}
install.packages("devtools")
install_github("kablag/RDML")
```

0. Start libraries.
```{r}
setwd('c:/Users/admin/SkyDrive/DimaTools/')
library(devtools)
library(RDML)
```

1. Import RDML file and print its overview as dendrogram, check correct target naming.
```{r}
q = RDML$new(file.choose()) #will open file selection dialog 
q$AsDendrogram() #most valuable picture
```

2. Make data table.
Added columns:
cq - cq as calculated by a qPCR machine software;
trgt_delta – cq difference for paired wells;
smaller.one – the one of paired wells with smaller cq (Boolean);
targetstr – target name obtained from the well sample name as set in a qPCR machine software (should be written “Target_State”, note underscore);
state – state name (whatever is detected by the allele-specific primer, for example SNP allele) obtained from the well sample name as set in a qPCR machine software (should be written “Target_State”, note underscore), if no state found – ‘Pos’ would be set as state;
pair_well – pair well deduced from sample and target names;
pair_state – the well state as described above, but for the paired well;
ctrl_ok – check qPCR for control gene and compare the paired wells control qPCR results (Boolean);
ctrl_ok_paired – sums check for both paired wells (Boolean);
ctrl_delta – cq difference for control qPCR reaction for the paired wells;
outlier – checks for plausible evaporation or other control qPCR problems in the wells by calculating outliers in the control gene qPCR endpoint fluorescence (the ‘endpoint’ column in the table) (Boolean);
outlier_paired – sums outlier check for the paired wells (Boolean);
rxn_ok – sums quality controls for both wells: control reactions, plausible evaporation and control gene cq difference (Boolean);
rxn_call – allele call for both wells, it shows the sample states for the target; won’t show the call if ‘rxn_ok’ is false either for the well or for the paired well;
endpoint – florescence data for the last cycle of the reaction.
```{r}
qtab = q$AsTable(add.columns = list(cq = data$cq, trgt_delta = NA, smaller.one = NA, targetstr = sapply(strsplit(data$tar$id,'_'),'[',1), state = sapply(strsplit(data$tar$id,'_'),'[',2), pair_well = NA, pair_state = NA, ctrl_ok = T, ctrl_ok_paired = T, ctrl_delta = NA, outlier = F, outlier_paired = F, rxn_ok = F, rxn_call = NA))
fdata = q$GetFData(qtab, long.table = T)
qtab$endpoint = fdata$fluor[fdata$cyc == max(fdata$cyc)]
```

3. Set control gene name. Check target and sample names, target states.
```{r}
ctrlgene = 'B2m' # control gene name, as written in the data file
ctrl.delta = 1.5 # allowed Cq differnece for ctrl gene for paired wells
trgt.delta = 2 # Cq difference for allele call for paired wells (for close Cq values)
cq.threshold = 30 # Cq for negative allele call
targets = unique(qtab$targetstr) # get target names from unknown wells
samples = unique(qtab$sample[qtab$sample.type == 'unkn']) # get sample names from unknown wells
ctrl.plz = unique(qtab$sample[qtab$sample.type == 'pos']) # get ctrl plazmids names from wells marked as positive control
positions = unique(qtab$position) #get occupied wells
#print obtained values
cat('Targets\n', targets, '\n')
targets = targets[targets != ctrlgene]
cat('Target states\n', unique(qtab$state), '\n')
cat('Samples\n', samples, '\n')
cat('Control gene set to', ctrlgene, '\n')
cat('Control plazmids\n', ctrl.plz)
```

4. Updating table (well pairs, bad ctrls, possible evaporation)
```{r}
# Add paired well to qtab$pair_well 
for (position in positions){
  pair = qtab$targetstr == qtab$targetstr[qtab$position == position & qtab$targetstr != ctrlgene][1] & qtab$sample == qtab$sample[qtab$position == position][1] & qtab$position != position 
  qtab$pair_well[qtab$position == position] = qtab$position[pair]
}

# qtab$ctrl_ok = T #fast reset
# qtab$ctrl_delta = NA #fast reset

# Ctrl check and ctrl Cq delta for paired wells (qtab$ctrl_delta) 
# Trgt gene Cq delta (qtab$trgt_delta)
for (position in positions){
  ctrl.gene.cq = qtab$cq[qtab$position == position & qtab$target == ctrlgene]
  genes.cq = qtab$cq[qtab$position == position]
  genes.cq.pair.well = qtab$cq[qtab$pair_well == position]
  qtab$trgt_delta[qtab$position == position] = round(abs(genes.cq - genes.cq.pair.well), digits = 3)
  if (ctrl.gene.cq > cq.threshold || is.na(ctrl.gene.cq)) {
    qtab$ctrl_ok[qtab$position == position] = F
    qtab$ctrl_ok_paired[qtab$position == position] = F
    qtab$ctrl_ok_paired[qtab$pair_well == position] = F
  } else {
    qtab$ctrl_delta[qtab$position == position] = round(abs(ctrl.gene.cq - qtab$cq[qtab$pair_well == position & qtab$target == ctrlgene]), digits = 3)
  }
}

# qtab$smaller.one column
for (position in positions){
  genes.cq = qtab$cq[qtab$position == position]
  genes.cq.pair.well = qtab$cq[qtab$pair_well == position]
  qtab$smaller.one[qtab$position == position] = ifelse(!is.na(genes.cq < genes.cq.pair.well), genes.cq < genes.cq.pair.well, ifelse(is.na(genes.cq), F, T))
}

# Check for evaporation by outlier control gene values, fill in 'qtab$outlier' and 'qtab$outlier_paired' columns.
outlier_pos = qtab$position[qtab$endpoint %in% boxplot.stats(qtab$endpoint[qtab$target == ctrlgene & qtab$ctrl_ok == T])$out]
qtab$outlier[qtab$position %in% outlier_pos] = T
qtab$outlier_paired[qtab$position %in% outlier_pos | qtab$pair_well %in% outlier_pos] = T

# qtab$rxn_ok
qtab$rxn_ok[qtab$ctrl_ok_paired == T & qtab$outlier_paired == F & qtab$ctrl_delta < ctrl.delta] = T

# qtab$state NAs substitution
qtab$state[is.na(qtab$state) & qtab$cq < cq.threshold & !is.na(qtab$cq) & qtab$rxn_ok == T] = 'Pos'
qtab$state[is.na(qtab$state) & (qtab$cq > cq.threshold | is.na(qtab$cq)) & qtab$rxn_ok == T] = 'Neg'

# qtab$pair_state
for (position in positions){
  qtab$pair_state[qtab$position == position] = qtab$state[qtab$pair_well == position]
}

# output for quality check: low ctrl gene Cq and evaporation prediction (based on endpoint fluorecence outliers for ctrl gene)
neg_ctrlgene_sample_wells = unique(qtab$position[qtab$ctrl_ok == F & qtab$sample.type == 'unkn'])
cat('Sample wells with negative', ctrlgene, ':\n', ifelse(length(neg_ctrlgene_sample_wells) == 0,'None', neg_ctrlgene_sample_wells), '\n')
cat('Possible evaporation in\n', unique(qtab$position[qtab$outlier == T])) 
```

5. Fill in qtab$rxn_call
```{r}
qtab$rxn_call = NA
# qtab$rxn_call fill in
for (position in positions) {
  for (target in qtab$targetstr[qtab$position == position]){
    index = qtab$position == position & qtab$targetstr == target
    homozig = qtab$rxn_ok[index] == T & (qtab$trgt_delta[index] > trgt.delta | is.na(qtab$trgt_delta[index])) & qtab$smaller.one[index] == T
    hetero = qtab$rxn_ok[index] == T & qtab$trgt_delta[index] < trgt.delta & !is.na(qtab$trgt_delta[index]) 
    if (homozig & qtab$cq[index] < cq.threshold & !is.na(qtab$cq[index])){
      qtab$rxn_call[index] = paste(qtab$state[index],qtab$state[index], sep = '/')
      qtab$rxn_call[qtab$pair_well == position & qtab$targetstr == target] = paste(qtab$state[index],qtab$state[index], sep = '/')
    }
    if (hetero & qtab$cq[index] < cq.threshold & !is.na(qtab$cq[index])){
      qtab$rxn_call[index] = paste(qtab$state[index], qtab$pair_state[index], sep = '/')
      qtab$rxn_call[qtab$pair_well == position & qtab$targetstr == target] = paste(qtab$state[index], qtab$pair_state [index], sep = '/')
    }
  }
}
```

6. output...
